src/TopSurface.cpp

```diff
@@ -0,0 +1,64 @@
+//Copyright (c) 2017 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#include "infill.h"
+#include "LayerPlan.h"
+#include "TopSurface.h"
+
+namespace cura
+{
+
+TopSurface::TopSurface()
+{
+    //Do nothing. Areas stays empty.
+}
+
+TopSurface::TopSurface(SliceMeshStorage& mesh, size_t layer_number, size_t part_number)
+{
+    //The top surface is all parts of the mesh where there's no mesh above it, so find the layer above it first.
+    Polygons mesh_above;
+    if (layer_number < mesh.layers.size() - 1)
+    {
+        mesh_above = mesh.layers[layer_number + 1].parts[part_number].print_outline;
+    } //If this is the top-most layer, mesh_above stays empty.
+
+    Polygons mesh_this = mesh.layers[layer_number].getOutlines();
+    areas = mesh_this.difference(mesh_above);
+}
+
+bool TopSurface::sand(const SliceMeshStorage& mesh, const GCodePathConfig& line_config, LayerPlan& layer)
+{
+    if (areas.empty())
+    {
+        return false; //Nothing to do.
+    }
+    //Generate the lines to cover the surface.
+    const EFillMethod pattern = mesh.getSettingAsFillMethod("sanding_pattern");
```

Comment:
Yes. Most of our patterns leave holes open. We have 3 patterns that don't: Lines, ZigZag and Concentric. Other patterns, like triangles, would leave holes open at 100% infill (line distance equal to line width) and that's not expected.

As the ironing was supposed to make travel moves, there would be no real difference between lines and zigzag. Later on I added a flow setting which didn't make it travel moves any more, so there would be a difference, but these travel moves are between adjacent lines and since the default line distance is at 0.1mm, that travel move is always very small anyway, so I just went for ZigZag.